system_metrics/basic_sender.go

```diff
@@ -174,6 +182,161 @@ func (s *SystemMetricsClient) AddSystemMetrics(ctx context.Context, id string, d
 	return err
 }
 
+// SystemMetricsWriter is a wrapper around a stream of system metrics data
+// that implements buffering with timed flushes and the WriteCloser interface.
+type SystemMetricsWriter struct {
+	mu            sync.Mutex
+	ctx           context.Context
+	cancel        context.CancelFunc
+	catcher       grip.Catcher
+	id            string
+	stream        internal.CedarSystemMetrics_StreamSystemMetricsClient
+	buffer        []byte
+	maxBufferSize int
+	lastFlush     time.Time
+	timer         *time.Timer
+	flushInterval time.Duration
+	closed        bool
+}
+
+// Write adds provided data to the buffer, flushing it if it exceeds its max
+// size. Write will fail if the writer was previously closed or if a timed flush
+// encountered an error.
+func (s *SystemMetricsWriter) Write(data []byte) (n int, err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.catcher.HasErrors() {
+		return 0, errors.Wrapf(s.catcher.Resolve(), "writer already closed due to error")
+	}
+	if s.closed {
+		return 0, errors.New("writer already closed")
+	}
+
+	s.buffer = append(s.buffer, data...)
+	if len(s.buffer) > s.maxBufferSize {
+		if err = s.flush(); err != nil {
+			s.catcher.Add(err)
+			s.catcher.Add(s.close())
+			return 0, errors.Wrapf(s.catcher.Resolve(), "problem writing data")
+		}
+	}
+	return len(data), nil
+}
+
+// Close flushes any remaining data in the buffer and closes the stream. If
+// the writer was previously closed, this returns an error.
+func (s *SystemMetricsWriter) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.catcher.HasErrors() {
+		return errors.Wrapf(s.catcher.Resolve(), "writer already closed due to error")
+	}
+	if s.closed {
+		return errors.New("writer already closed")
+	}
+
+	s.catcher.Add(s.flush())
+	s.catcher.Add(s.close())
+	return s.catcher.Resolve()
+}
+
+func (s *SystemMetricsWriter) flush() error {
+	data := &internal.SystemMetricsData{
+		Id:   s.id,
+		Data: s.buffer,
+	}
+
+	if err := s.stream.Send(data); err != nil {
+		return errors.Wrapf(err, "problem sending data for id %s", s.id)
+	}
+
+	s.buffer = []byte{}
+	return nil
+}
+
+func (s *SystemMetricsWriter) close() error {
+	s.cancel()
+	s.closed = true
+	_, err := s.stream.CloseAndRecv()
+	return err
+}
+
+func (s *SystemMetricsWriter) timedFlush() {
+	s.mu.Lock()
+	s.timer = time.NewTimer(s.flushInterval)
+	s.mu.Unlock()
+	defer s.timer.Stop()
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-s.timer.C:
+			s.mu.Lock()
+			if len(s.buffer) > 0 && time.Since(s.lastFlush) >= s.flushInterval {
+				if err := s.flush(); err != nil {
+					s.catcher.Add(errors.Wrapf(err, "problem with timed flush for id %s", s.id))
+					s.catcher.Add(s.close())
+				}
+			}
+			_ = s.timer.Reset(s.flushInterval)
+			s.mu.Unlock()
```

Comment:
Instead of putting the unlock far from the lock statement, can we just wrap this case in a `func() { ... }()` block and use `s.mu.Lock()` followed by defer `s.mu.Unlock()`?